lib/path/default.nix

```diff
@@ -149,6 +209,100 @@ in /* No rec! Add dependencies on this file at the top. */ {
+  Throws an error if the paths don't share the same filesystem root.
+
+  Laws:
+
+  - Equivalent to whether some subpath exists that can be appended to the first path to get the second path:
+
+        hasPrefix p q <-> exists s . append p s == q
+
+  - `hasPrefix` is a [non-strict partial order](https://en.wikipedia.org/wiki/Partially_ordered_set#Non-strict_partial_order) over the set of all path values
+
+  - `lib.path.hasProperPrefix` is the [corresponding strict partial order](https://en.wikipedia.org/wiki/Partially_ordered_set#Correspondence_of_strict_and_non-strict_partial_order_relations):
+
+        hasPrefix p q <-> hasProperPrefix p q || p == q
+
+  Type:
+    hasPrefix :: Path -> Path -> Bool
+
+  Example:
+    hasPrefix /foo /foo/bar
+    => true
+    hasPrefix /foo /foo
+    => true
+    hasPrefix /foo/bar /foo
+    => false
+    hasPrefix /. /foo
+    => true
+  */
+  hasPrefix = withTwoDeconstructedPaths "lib.path.hasPrefix" (prefix: deconPrefix: path: deconPath:
+    take (length deconPrefix.components) deconPath.components == deconPrefix.components
+  );
+
+  /*
+  Whether the second path is a prefix of the first path and not equal to it.
+  Throws an error if the paths don't share the same filesystem root.
+
+  Laws:
+
+  - `hasProperPrefix` is a [strict partial order](https://en.wikipedia.org/wiki/Partially_ordered_set#Strict_partial_order) over the set of all path values
+
+  - `lib.path.hasPrefix` is the [corresponding non-strict partial order](https://en.wikipedia.org/wiki/Partially_ordered_set#Correspondence_of_strict_and_non-strict_partial_order_relations):
+
+        hasProperPrefix p q <-> hasPrefix p q && p != q
+
+  Type:
+    hasProperPrefix :: Path -> Path -> Bool
+
+  Example:
+    hasProperPrefix /foo /foo/bar
+    => true
+    hasProperPrefix /foo /foo
+    => false
+    hasProperPrefix /foo/bar /foo
+    => false
+    hasProperPrefix /. /foo
+    => true
+  */
+  hasProperPrefix = withTwoDeconstructedPaths "lib.path.hasProperPrefix" (prefix: deconPrefix: path: deconPath:
+    take (length deconPrefix.components) deconPath.components == deconPrefix.components
+    && prefix != path
+  );
+
+  /*
+  Remove a first prefix path from a second path, the result is a normalised subpath string, see `lib.path.subpath.normalise`.
+  Throws an error if the second path doesn't start with the prefix, see `lib.path.hasPrefix`.
+  Throws an error if the first and the second path don't share the same filesystem root.
+
+  Laws:
+
+  - Inverses `append` for normalised subpaths:
+
+      removePrefix p (append p s) == subpath.normalise s
+
+  Type:
+    removePrefix :: Path -> Path -> String
+
+  Example:
+    removePrefix /foo /foo/bar
+    => "./bar"
+    removePrefix /foo /foo
+    => "./."
+    removePrefix /foo/bar /foo
+    => <error>
+    removePrefix /. /foo
+    => "./foo"
+  */
+  removePrefix = withTwoDeconstructedPaths "lib.path.removePrefix" (prefix: deconPrefix: path: deconPath:
```

Comment:
I now added a third function to this PR, `removePrefix` because it's very related to the other two and can share a lot of the code. Previously in https://github.com/NixOS/nixpkgs/pull/200718 I called it `relativeTo`